landoapi/api/treestatus.py

```diff
@@ -0,0 +1,619 @@
+
+    This also shows status, reason and tags from last Tree Log.
+    """
+    result = copy.deepcopy(DEFAULT_TREE)
+    result.update(tree.to_dict())
+
+    if tags is not None:
+        result["tags"] = json.loads(tags)
+
+    if status is not None:
+        result["status"] = status
+
+    if reason is not None:
+        result["reason"] = reason
+
+    if log_id is not None:
+        result["log_id"] = log_id
+
+    result["model"] = tree
+
+    return CombinedTree(**result)
+
+
+def combinedtree_as_dict(tree: CombinedTree) -> dict[str, Any]:
+    """Convert a `CombinedTree` to a dict.
+
+    Removes the `model` field as part of the conversion.
+    """
+    return {field: value for field, value in tree._asdict().items() if field != "model"}
+
+
+def result_object_wrap(f: Callable) -> Callable:
+    """Wrap the value returned from `f` in a result object.
+
+    Return a result wrapped in an object with a `result` key, like so:
+        {"result": ...}
+    """
+
+    @functools.wraps(f)
+    def wrap_output(*args, **kwargs) -> dict[str, Any]:
+        result = f(*args, **kwargs)
+        return {"result": result}
+
+    return wrap_output
+
+
+def serialize_last_state(old_tree: dict, new_tree: CombinedTree) -> str:
+    """Serialize a `last_state` value for a `StatusChangeTree`."""
+    return json.dumps(
+        {
+            "status": old_tree["status"],
+            "reason": old_tree["reason"],
+            "tags": old_tree["tags"],
+            "log_id": old_tree["log_id"],
+            "current_status": new_tree.status,
+            "current_reason": new_tree.reason,
+            "current_tags": new_tree.tags,
+            "current_log_id": new_tree.log_id,
+        }
+    )
+
+
+@cache.memoize()
+def get_tree_by_name(tree: str) -> Optional[CombinedTree]:
+    """Retrieve a `CombinedTree` representation of a tree by name.
+
+    Returns `None` if no tree can be found.
+    """
+    session = db.session
+
+    query = (
+        session.query(Tree)
+        .distinct(Tree.tree)
+        .add_columns(
+            Log._tags,
+            Log.status,
+            Log.reason,
+            Log.id,
+        )
+        .outerjoin(
+            Log,
+            Log.tree == Tree.tree,
+        )
+        .order_by(Tree.tree.desc(), Log.when.desc())
+        .filter(Tree.tree == tree)
+    )
+
+    result = query.one_or_none()
+    if result:
+        return get_combined_tree(*result)
+
+
+def remove_tree_by_name(tree_name: str):
+    """Completely remove the tree with the specified name.
+
+    Note: this function commits the session.
+    """
+    session = db.session
+    tree = session.query(Tree).filter_by(tree=tree_name).one_or_none()
+    if not tree:
+        raise ProblemException(
+            404,
+            f"No tree {tree_name} found.",
+            "The tree does not exist.",
+            type="https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/404",
+        )
+    session.delete(tree)
```

Comment:
May be better to implement some type of archiving functionality, to avoid accidental deletions of trees and records (e.g., have a `Tree.is_archived` flag.)